Add sharded repodata documentation (#16267)
conda/conda: docs/source/dev-guide/deep-dives/sharded.md

# Sharded repodata

This document provides an overview on how `conda` implements
[CEP-16 Sharded Repodata](https://conda.org/learn/ceps/cep-0016).

Sharded repodata splits `repodata.json` into an index mapping package names to
shard hashes in `repodata_shards.msgpack.zst`. A shard contains repodata for
every package with a given name. Since shards are named after a hash of their
contents, they can be cached without having to check the server for freshness.
Individual shards only need to change when an individual package has changed, so
only the much smaller index has to be re-fetched often.

## Sharded Repodata in conda

> For `conda` and `conda-libmamba-solver`, which contained the sharsd
> implementation later ported to `conda`, we wanted a way to implement sharded

Reviewer: ```suggestion
Originally developed in `conda-libmamba-solver` and later ported into `conda`, we wanted a way to implement sharded
```

We have two separate thread workers for fetching cache and network data. These
threads communicate to each other via the following queues:

- **cache_in_queue** every requested shard goes here first where the cache
  worker sees if we have a valid cache record.
- **cache_miss_queue** for every shard not in cache, we send it this queue where
  the network worker thread downloads it.
- **shard_out_queue** once a shard has been fetched from either the cache or
  network worker threads, it is placed here so we can gather all needed
  shards at the end to build our repodata subset.

:::{mermaid}

    sequenceDiagram
        loop
            Main ->> Main: "Fetch" in-memory shard
            Main ->> Cache: Fetch shard
            Cache ->> Network: Cache miss
            Cache ->> Main: Cache hit
            Network ->> Main: Network result
            Main ->> Main: Find new (channel, package) from shard data
        end

:::

## Source code

The shard handling code is split into `conda/_private/shards/shards.py`, `conda/_private/shards/cache.py`,
`conda/_private/shards/subset.py`, `conda/_private/shards/typing.py`, and `conda/_private/shards/misc.py` in `conda/_private/`.

> Additional code in `conda/_private/index.py` calls
> `build_repodata_subset()` and converts the resulting repodata to `libmamba`
> objects.

Reviewer: ```suggestion
`conda/gateways/shards/` re-exports `build_repodata_subset()`. When `context.repodata_use_shards` is enabled, `conda/plugins/manager.py` injects it into solver backends that accept a `build_repodata_subset` constructor parameter.
Solver plugins such as `conda-libmamba-solver` pass the injected callable to their index helper, which calls it and converts the resulting repodata to solver objects.
```

type of each channel.

### `conda/_private/shards/cache.py`

`conda/_private/shards/cache.py` implements a sqlite3 cache used to store individual shards.
When traversing shards, the cache is checked before making a network request.
The shards cache is a single database for all channels in
`$CONDA_PREFIX/pkgs/cache/repodata_shards.db`.

The shards index `repodata_shards.msgpack.zst` is cached in the same way as
`repodata.json`, in individual files in `$CONDA_PREFIX/pkgs/cache/` named after
URL hashes. A `has_<format>` remembers if a channel has shards, or not. If
`has_shards` is `false` then we wait 7 days after `last_checked` to make another
request looking for `repodata_shards.msgpack.zst`. The same system remembers
whether a channel provides `repodata.json.zst`, and stores `ETag` and
`Last-Modified` used to refresh the cache.

```
...
"has_shards": {
    "last_checked": "2025-10-15T17:19:44.408989Z",
    "value": true
},
```

### `conda/_private/shards/typing.py`

`conda/_private/shards/typing.py` provides type hints for data structures used in sharded
repodata, but it is not normative; it only includes fields used by the sharded
repodata system.

> ### `tests/test_shards.py`

Reviewer: ```suggestion
### `tests/shards/test_shards.py`
```

The sharded repodata tests maintain 100% code coverage of the shards-related code
in `conda/_private/shards/*.py`.

## Example dependency graph for Python

This is what Python's dependencies look like on `conda-forge` as of this writing.

If sharded repodata is asked to install Python, we look for `python` in every
active channel. The `python` shard(s) tells us we can fetch `bzip2`, `libffi`,
`...` in parallel, discovering a third layer including `icu`, `ca-certificates`,
and others. `ca-certificates` also depends on some virtual packages, but the

> traversal quickly determine that these packages don't appear in any channel by

Reviewer: ```suggestion
traversal quickly determines that these packages don't appear in any channel by
```
